lib/Dialect/FIRRTL/Export/FIREmitter.cpp

```diff
@@ -0,0 +1,643 @@
+  /// has not yet been emitted or it was emitted inline.
+  Optional<StringRef> lookupEmittedName(Value value) {
+    auto it = valueNames.find(value);
+    if (it != valueNames.end())
+      return {it->second};
+    return {};
+  }
+
+private:
+  /// The stream we are emitting into.
+  llvm::raw_ostream &os;
+
+  /// Whether we have encountered any errors during emission.
+  bool encounteredError = false;
+
+  /// Current level of indentation. See `indent()` and
+  /// `addIndent()`/`reduceIndent()`.
+  unsigned currentIndent = 0;
+
+  /// The names used to emit values already encountered. Anything that gets a
+  /// name in the output FIR is listed here, such that future expressions can
+  /// reference it.
+  DenseMap<Value, StringRef> valueNames;
+  StringSet<> valueNamesStorage;
+
+  void addValueName(Value value, StringAttr attr) {
+    valueNames.insert({value, attr.getValue()});
+  }
+  void addValueName(Value value, StringRef str) {
+    auto it = valueNamesStorage.insert(str);
+    valueNames.insert({value, it.first->getKey()});
+  }
+};
+} // namespace
+
+LogicalResult Emitter::finalize() { return failure(encounteredError); }
+
+/// Emit an entire circuit.
+void Emitter::emitCircuit(CircuitOp op) {
+  indent() << "circuit " << op.name() << " :\n";
+  addIndent();
+  for (auto &bodyOp : *op.getBody()) {
+    if (encounteredError)
+      return;
+    TypeSwitch<Operation *>(&bodyOp)
+        .Case<FModuleOp, FExtModuleOp>([&](auto op) {
+          emitModule(op);
+          os << "\n";
+        })
+        .Default([&](auto op) {
+          emitOpError(op, "not supported for emission inside circuit");
+        });
+  }
+  reduceIndent();
+}
+
+/// Emit an entire module.
+void Emitter::emitModule(FModuleOp op) {
+  indent() << "module " << op.getName() << " :\n";
+  addIndent();
+
+  // Emit the ports.
+  auto ports = op.getPorts();
+  emitModulePorts(ports, op.getArguments());
+  if (!ports.empty() && !op.getBodyBlock()->empty())
+    os << "\n";
+
+  // Emit the module body.
+  emitStatementsInBlock(*op.getBodyBlock());
+
+  reduceIndent();
+  valueNames.clear();
+  valueNamesStorage.clear();
+}
+
+/// Emit an external module.
+void Emitter::emitModule(FExtModuleOp op) {
+  indent() << "extmodule " << op.getName() << " :\n";
+  addIndent();
+
+  // Emit the ports.
+  auto ports = op.getPorts();
+  emitModulePorts(ports);
+
+  // Emit the optional `defname`.
+  if (op.defname() && !op.defname()->empty())
+    indent() << "defname = " << op.defname() << "\n";
+
+  // Emit the parameters.
+  if (auto params = op.parameters()) {
+    for (auto param : *params) {
+      indent() << "parameter " << param.first << " = ";
+      TypeSwitch<Attribute>(param.second)
+          .Case<IntegerAttr>([&](auto attr) { os << attr.getValue(); })
+          .Case<FloatAttr>([&](auto attr) {
+            SmallString<16> str;
+            attr.getValue().toString(str);
+            os << str;
```

Comment:
Is this different than `os << attr;` ?

Reply:
The attribute printer would generally include the type of the attribute, so `3.141 : f64` instead of just `3.141`. I tried printing `attr.getValue()` directly, but there doesn't seem to be an impl of `<<` for that.